src/sessions/p2p_session.rs

```diff
@@ -842,4 +861,60 @@ impl<T: Config> P2PSession<T> {
             self.event_queue.pop_front();
         }
     }
+
+    fn compare_local_checksums_against_peers(&mut self) {
+        match self.desync_detection {
+            DesyncDetection::On { interval } => {
+                if self.current_frame() % interval as i32 != 0 {
+                    return;
+                }
+
+                for (_, remote) in &self.player_reg.remotes {
+                    for (remote_frame, remote_checksum) in remote.checksum_history() {
+                        if let Some(local_checksum) = self.checksum_history.get(remote_frame) {
+                            if *local_checksum != *remote_checksum {
+                                self.event_queue.push_back(GGRSEvent::DesyncDetected {
+                                    frame: *remote_frame,
+                                    local_checksum: *local_checksum,
+                                    remote_checksum: *remote_checksum,
+                                    addr: remote.peer_addr(),
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+            DesyncDetection::Off => (),
+        }
+    }
+
+    fn check_checksum_send_interval(&mut self) {
+        match self.desync_detection {
+            DesyncDetection::On { interval } => {
+                let frame_to_send = self.sync_layer.last_saved_frame() - 1;
+                let current = self.current_frame();
+                
+                if current % interval as i32 == 0 && frame_to_send > self.max_prediction as i32 {
+                    let cell = self
+                        .sync_layer
+                        .saved_state_by_frame(frame_to_send)
+                        .expect(&format!("cell not found!: frame {}", frame_to_send));
+
+                    if let Some(checksum) = cell.checksum() {
+                        for (_, remote) in &mut self.player_reg.remotes {
+                            remote.send_checksum_report(frame_to_send, checksum);
+                        }
+                        // collect locally for later comparison
+                        self.checksum_history.insert(frame_to_send, checksum);
+                    }
+                }
+                while self.checksum_history.len() > MAX_CHECKSUM_HISTORY_SIZE {
```

Comment:
similar to the comment above, why does this have to be a while-loop? It should not be.